XMonad/Layout/FixedAspectRatio.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Layout.FixedAspectRatio
+-- Copyright   :  (c) Yecine Megdiche <yecine.megdiche@gmail.com>
+-- License     :  BSD3-style (see LICENSE)
+--
+-- Maintainer  :  Yecine Megdiche <yecine.megdiche@gmail.com>
+-- Stability   :  unstable
+-- Portability :  unportable
+--
+-- Layout modifier for user provided per-window aspect ratios.
+--
+-----------------------------------------------------------------------------
+
+module XMonad.Layout.FixedAspectRatio
+  (
+    -- * Usage
+    -- $usage
+    fixedAspectRatio
+  , FixedAspectRatio
+  , ManageAspectRatio(..)
+  ) where
+
+
+import           Control.Arrow
+import qualified Data.Map                      as M
+import           Data.Ratio
+
+import           XMonad
+import           XMonad.Layout.Decoration
+import           XMonad.Layout.LayoutHints
+
+-- $usage
+-- You can use this module with the following in your @~\/.xmonad\/xmonad.hs@:
+--
+-- > import XMonad.Layout.FixedAspectRatio
+-- Then add it to your layout:
+--
+-- > myLayout = fixedAspectRatio (0.5, 0.5) $ Tall 1 (3/100) (1/2)  ||| Full ||| etc..
+-- > main = xmonad def { layoutHook = myLayout }
+--
+-- Which will center the (eventually) shrinked windows in their assigned
+-- rectangle.
+--
+-- For a layout modifier that automatically sets the aspect ratio
+-- depending on the size hints (for example for programs like mpv),
+-- see "XMonad.Layout.LayoutHints"
+--
+-- See "XMonad.Doc.Extending#Editing_the_layout_hook" for more info on
+-- the 'layoutHook'.
+--
+-- You also want to add keybindings to set and clear the aspect ratio:
+--
+-- >      -- Set the aspect ratio of the focused window to 16:9
+-- >   ,((modm, xK_a), withFocused $ sendMessage . Fix (16 / 9))
+-- >
+-- >      -- Clear the aspect ratio from the focused window
+-- >   ,((modm .|. shiftMask, xK_a), withFocused $ sendMessage . Reset)
+--
+-- There's one caveat: to keep the usage of the modifier simple, it
+-- doesn't remove a window from its cache automatically. Which means
```

Comment:
I was going to suggest to deal with this but I noticed this problem plagues many other modules as well, even ones like BoringWindows that should really know better. :-(
(I remember fixing a couple memory leaks back in 2009/2010 when I started using xmonad, but I don't use any of these modules so I didn't bother. It might still be a problem, but other people have different usage patterns, higher tolerance to breakage, and likely also smaller uptimes.)

On the other hand it seems that something like this may take care of the garbage collection:

```haskell
  modifyLayoutWithUpdate (FixedAspectRatio ratios placement) w r = do
    let ws = W.integrate' (W.stack w)
    let ratios' = M.filterWithKey (\k _ -> k `elem` ws) ratios
    let m' = FixedAspectRatio ratios' placement
    (, Just m') <$> runLayout w r
```

It may be too eager (e.g. when used as a SubLayout, not that I can imagine a scenario when that would make sense), so this might be a safer (and more complicated):

```haskell
  modifyLayoutWithUpdate (FixedAspectRatio ratios placement) w r = do
    ws <- gets (W.index . W.view (W.tag w) . windowset)
    …
```

Reply:
And I suppose layouts have no way of knowing when windows are killed since they don't see the events? 

I could fix this here, but since this is a problem, why don't we think of a better alternative? I was thinking of adding a combinator that installs an handleEvent that simply broadcasts a message (for example `WindowKilled w`). That way we can patch layouts easily. Would this work? And which event should we listen to?

Reply:
Layouts see only [events that aren't handled by the core](https://github.com/xmonad/xmonad/blob/4a0b166998a838d91f002991da82507a2d04d3bb/src/XMonad/Main.hs#L426) but [DestroyWindowEvent obviously is handled](https://github.com/xmonad/xmonad/blob/4a0b166998a838d91f002991da82507a2d04d3bb/src/XMonad/Main.hs#L315) so layouts don't see this one.

Reply:
I've opened https://github.com/xmonad/xmonad/pull/284 to broadcast `DestroyWindowEvent`. If we decide against it, then I'll add one of your suggestions.

Reply:
Yeah I've seen that, I just wanted to focus on the 0.17 stuff so I'd rather not dig deep into this just yet.